em-app/examples/harmonize/src/main.rs

```diff
@@ -0,0 +1,174 @@
+    if args.len() < 2 {
+        return Err(format!("Usage: ftxsgx-runner {} app-config-id", args[0]));
+    }
+
+    // Generate key - mbedtls::Pk has required trait implemented - customers may choose any other mechanism to create certificates
+    let mut rng = FtxRng;
+    let mut key = Pk::generate_rsa(&mut rng, 3072, 0x10001).unwrap();
+
+    // This is always on localhost but the port might be different
+    let node_agent_url = "http://localhost:9092";
+
+    // This must be on localhost otherwise local attestation will not work
+    let em_server = "ccm.fortanix.com";
+    let em_port = 443;
+
+    let em_ca_cert = Some(Arc::new(Certificate::from_pem_multiple(concat!(include_str!("../certs/em_ca_cert.pem"), "\0").as_bytes()).map_err(|e| format!("Invalid CA Cert for CCM: {:?}", e))?));
+    let em_crl = None;
+    let sdkms_ca_cert = em_ca_cert.clone();
+    let sdkms_crl = None;
+
+    let storage_ca = Some(Arc::new(Certificate::from_pem_multiple(concat!(include_str!("../certs/aws_s3.pem"), "\0").as_bytes()).map_err(|e| format!("Invalid CA Cert for CCM: {:?}", e))?));
+    let storage_crl = None;
+    
+    let config_id = args[1].clone();
+
+    let result = get_certificate(node_agent_url,"localhost", &mut key, None, Some(&config_id)).map_err(|e| format!("Error: {}", e))?;
+    println!("\na. Public certificate is signed: \n{}", serde_json::to_string_pretty(&result.certificate_response).map_err(|e| format!("Failed decoding certificate response: {:?}", e))?);
+
+    let mut cert_pem = result.certificate_response.certificate.ok_or("Missing certificate in response")?;
+    cert_pem.push('\0');
+
+    let app_cert = Arc::new(Certificate::from_pem_multiple(&cert_pem.as_bytes()).map_err(|e| format!("Parsing certificate failed: {:?}", e))?);
+    let key = Arc::new(key);
+
+    let config = get_runtime_configuration(em_server, em_port, app_cert.clone(), key.clone(), em_ca_cert, em_crl).map_err(|e| format!("Error in client: {:?}", e))?;
+    println!("\nb. Application configuration: \n{}", serde_json::to_string_pretty(&config).map_err(|e| format!("Failed decoding response: {:?}", e))?);
+
+    let (input, input_credentials) = get_credentials("input", &config, app_cert.clone(), key.clone(), sdkms_ca_cert.clone(), sdkms_crl.clone())?;
+    println!("\nc.1. 'input' port dataset value from SDKMS: \n{}", serde_json::to_string_pretty(&input_credentials).unwrap());
+
+    let (output, output_credentials) = get_credentials("output", &config, app_cert.clone(), key.clone(), sdkms_ca_cert.clone(), sdkms_crl.clone())?;
+    println!("\nd.2. 'output' port dataset value from SDKMS: \n{}", serde_json::to_string_pretty(&output_credentials).unwrap());
+
+    let query_string = input_credentials.query_string.from_base64().map_err(|e| format!("Failed decoding query string: {:?}", e))?;
+    let query_string = &String::from_utf8(query_string).map_err(|e| format!("Query string is not utf-8: {:?}", e))?;
+    let url = input.location.to_owned() + "?" + query_string;
```

Comment:
Could use `Url::url` to construct this. Also, could validate that query_string really is a legal query string. Not a big deal here, but potentially becomes a concern in other code derived from the example.

Reply:
Updated so we go through Url::parse and raise errors. (Otherwise the 'set_query' url function does not throw errors)
The library also now only accepts Url structures.